### PR TITLE
cmd/server: concurrently search instead of serial iterations

### PR DESCRIPTION
**Pull request by adamdecaf**
This change will spread CPU usage across available cores and increase memory usage where possible, which offers better search performance.

These searches are still not nearly fast enough for me, but they are steady progress we can build from going forward.

```
// before (v0.17.1)
// Go 1.15.7
√ cmd/server % go test -bench BenchmarkSearch -run BenchmarkSearch -v
goos: darwin
goarch: amd64
pkg: github.com/moov-io/watchman/cmd/server
BenchmarkSearch__Addresses
BenchmarkSearch__Addresses-16      	      22	  50782847 ns/op
BenchmarkSearch__BISEntities
BenchmarkSearch__BISEntities-16    	     216	   5415917 ns/op
BenchmarkSearch__DPs
BenchmarkSearch__DPs-16            	    1855	    611191 ns/op
BenchmarkSearch__SDNs
BenchmarkSearch__SDNs-16           	      94	  12140552 ns/op
BenchmarkSearch__SSIs
BenchmarkSearch__SSIs-16           	     602	   1941815 ns/op
PASS
ok  	github.com/moov-io/watchman/cmd/server	15.228s
```
```
// after (sync.WaitGroup)
// Go 1.15.7
√ cmd/server % go test -bench BenchmarkSearch -run BenchmarkSearch -v
goos: darwin
goarch: amd64
pkg: github.com/moov-io/watchman/cmd/server
BenchmarkSearch__Addresses
BenchmarkSearch__Addresses-16      	      33	  34522864 ns/op
BenchmarkSearch__BISEntities
BenchmarkSearch__BISEntities-16    	    1161	   1021614 ns/op
BenchmarkSearch__DPs
BenchmarkSearch__DPs-16            	    3991	    296631 ns/op
BenchmarkSearch__SDNs
BenchmarkSearch__SDNs-16           	     394	   3042463 ns/op
BenchmarkSearch__SSIs
BenchmarkSearch__SSIs-16           	    2161	    505961 ns/op
PASS
ok  	github.com/moov-io/watchman/cmd/server	14.404s
```
```
// Go 1.16.3
√ cmd/server % go test -bench BenchmarkSearch -run BenchmarkSearch -v
go version
goos: darwin
goarch: amd64
pkg: github.com/moov-io/watchman/cmd/server
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkSearch__Addresses
BenchmarkSearch__Addresses-16      	      38	  27957405 ns/op
BenchmarkSearch__BISEntities
BenchmarkSearch__BISEntities-16    	    1180	   1014248 ns/op
BenchmarkSearch__DPs
BenchmarkSearch__DPs-16            	    4315	    277961 ns/op
BenchmarkSearch__SDNs
BenchmarkSearch__SDNs-16           	     392	   3023502 ns/op
BenchmarkSearch__SSIs
BenchmarkSearch__SSIs-16           	    2431	    503114 ns/op
PASS
ok  	github.com/moov-io/watchman/cmd/server	14.318s
```

### CODE REVIEW

File: cmd/server/search.go [hunk 1]
```diff
@@ -213,12 +223,19 @@ func (s *searcher) TopAltNames(limit int, alt string) []Alt {
 	}
 	xs := newLargest(limit)
 
+	var wg sync.WaitGroup
+	wg.Add(len(s.Alts))
```
Comment (atonks2) [inline cmd/server/search.go:227]: Should we add some kind of limit to avoid kicking off unknown quantities of goroutines?

Comment (adamdecaf) [reply cmd/server/search.go:227]: Yea and it is sorta handled by default. Goroutines are multiplexed across the available CPU's on the machine. This means they're not individual threads (in Java's case) and are a lot lighter weight.

Comment (adamdecaf) [reply cmd/server/search.go:227]: Typically `runtime.GOMAXPROCS(runtime.NUMCPU())` is used to limit this to the available CPU's, but I'm pretty sure that's the default now. 

https://github.com/letsencrypt/boulder/blob/5457680a9c8ce34d0456ccf289ed347a8529a31e/cmd/cert-checker/main.go#L335 is a common pattern to use

Comment (atonks2) [reply cmd/server/search.go:227]: Oh cool. I didn't know about the baked in limiting.

**Review by atonks2 — APPROVED** [review]

**Review by vxio — APPROVED** [review]
very nice!

curious if you have previous benchmark results on hand for comparison?

**Comment by adamdecaf** [comment]
Thanks! The first code block in the PR are the "before" results. We're seeing some significant improvements with the most naive concurrent approach.

**Pull request merged**